### DISCUSSION TITLE
Mapping savestates with keyboard keys

### DISCUSSION THREAD

**Discussion by mabakay** [BlitterStudio/amiberry#1558] [Q&A] [answered]
Is it possible to map Save State and Load State to keyboard keys like F5 and F9? Similarly with selecting a slot. Or maybe there's already a solution for this? The idea is to be able to press a key on the keyboard during gameplay instead of opening a window, going to the Save States tab, and using the mouse for all of that.

It would be easier to set up a station for a child where I can explain that this button is for this and that button is for that, rather than the child constantly having to call out – 'Dad, now I want to save and switch to another game' – or – 'Dad, now I want to play Superfrog where I left off.'

**Comment by midwan** [comment, marked as answer]
You can map the `Quick save state` and `Quick restore state` events to controller buttons:
![image](https://github.com/user-attachments/assets/8bb038f9-abf1-456b-8d23-5a52bfbafbb2)

There's no keyboard support for mapping custom events, but if you've got a controller with spare buttons, perhaps this will do the job.

**Reply by mabakay** [reply]
What is the difference between quick save state and regular save state? In which slot is the quick save saved and from which is loaded?

**Reply by midwan** [reply]
The regular one will open the GUI, to let you select the slot.
The Quick one, will just save to the relevant slot - there are normally 10 quick save state events, but only one is available to map to the controllers (the first one, so it will take the first slot).

I also found that there is a built-in mapping for Quick save states from the Keyboard, when the "Special" key + the numeric Keypad 0-9 and the Period key are pressed together. This is inherited from WinUAE, but I haven't tested it myself. I believe the "Special" combo is supposed to be `Shift-End`, so something like `Shift-End-Keypad 0-9` will save to the relevant slot.

**Reply by giantclambake** [reply]
>  I also found that there is a built-in mapping for Quick save states from the Keyboard, when the "Special" key + the numeric Keypad 0-9 and the Period key are pressed together. This is inherited from WinUAE, but I haven't tested it myself.

...**and** the Period key??... that sounds like cruel & unusual punishment ... ;^)

>  I believe the "Special" combo is supposed to be Shift-End, so something like Shift-End-Keypad 0-9 will save to the relevant slot.

Tested here, working as expected --> Shift-End-Keypad 8 creates a save state in Slot 8

Curio:  `Stunt Car Racer (1989)(Micro Style)-4_8.uss` //note: -4_8

To recreate: Load an existing `.uss` file [Stunt Car Racer (1989)(Micro Style)-4.uss], use above key-sequence [Shift-End-Keypad 8], resultant `.uss` file will be named as shown.

...not sure if that will be an issue or not, just reporting for sake of completeness... =)

Query: any similar key-sequence for Quick restore?

**Reply by mabakay** [reply]
Any idea what hotkey it will be on Pi400? In particular I'm referring to ASG v26 but that probably doesn't matter much.

**Reply by midwan** [reply]
There are a few different keyboard shortcuts inherited from WinUAE, that I haven't tested personally.
For example, `Shift + END + F6` will prompt a "Save save state" dialog, while `END + F6` will prompt a "Restore save state" one. 
`END + F1-F4` and `SHIFT + END + F1-F4` will do floppy drive insert/remove accordingly.
`END + F5` and `Shift + END + F5` will do the same for CD0.

I haven't tested these in Amiberry yet, but the mapping is there. The GUI interface is different, so it's possible some might be glitchy or not work, we'll have to test them. The keyboard mapping is all shown in the `osdep/keyboard.cpp` file. The processed events, are set up in `inputdevice.cpp`.

I don't see a keyboard shortcut for quick loading a savestate, from what I can see, but maybe it works with/without Shift, like the rest of them? Can't test at the moment to be sure. :)

**Reply by midwan** [reply]
Update: I see that the shortcuts for Save state dialog are not implemented yet in Amiberry, so those won't work (as you already know).
The floppy selection ones are there and should work. The CD one is also missing. Things to add in the TODO list :)

@mabakay It's still the END key on the PI400, though with the limited keyboard there, it might be hard to use that (I think it's a combo with the Fn key?). We need to have it mapped to a key that's not used inside Amiga emulation, so it's hard to use anything other than those extra keys.

**Reply by giantclambake** [reply]
> I haven't tested these in Amiberry yet, but the mapping is there. The GUI interface is different, so it's possible some might be glitchy or not work, we'll have to test them.

Sure.... I'm being a bit tunnel-visioned here ~ my main focus is on savestates atm...

`Shift + END + F6  ` will prompt a "Save save state" dialog
` END + F6  ` will prompt a "Restore save state" one

Neither of the above key-sequences work (with the GUI) ... I suspect that the triggers are working, as I notice mouse focus is effected, but the dialog windows don't appear (I doubt we have any GUI equivalents here?)

You want a separate ticket for this? (last time I checked, floppy insert/remove was working, but I'll recheck later)

**Reply by midwan** [reply]
I think this https://github.com/BlitterStudio/amiberry/issues/1559 will suffice :)

**Reply by giantclambake** [reply]
> I don't see a keyboard shortcut for quick loading a savestate, from what I can see, but maybe it works with/without Shift, like the rest of them? Can't test at the moment to be sure. :)

Doesn't seem to work without Shift ...maybe another qualifier?

```
src/include/savestate.h: extern void savestate_quick(int slot, int save);

src/savestate.cpp: void savestate_quick (int slot, int save)

src/inputdevice.cpp: 		savestate_quick ((code - AKS_STATESAVEQUICK) / 2, 1);
src/inputdevice.cpp: 		savestate_quick ((code - AKS_STATERESTOREQUICK) / 2, 0);

src/inputdevice.cpp: 	case AKS_STATERESTOREQUICK:
src/inputdevice.cpp: 	case AKS_STATERESTOREQUICK1:
src/inputdevice.cpp: 	case AKS_STATERESTOREQUICK2:
src/inputdevice.cpp: 	case AKS_STATERESTOREQUICK3:
src/inputdevice.cpp: 	case AKS_STATERESTOREQUICK4:
src/inputdevice.cpp: 	case AKS_STATERESTOREQUICK5:
src/inputdevice.cpp: 	case AKS_STATERESTOREQUICK6:
src/inputdevice.cpp: 	case AKS_STATERESTOREQUICK7:
src/inputdevice.cpp: 	case AKS_STATERESTOREQUICK8:
src/inputdevice.cpp: 	case AKS_STATERESTOREQUICK9:
src/inputdevice.cpp: 		savestate_quick ((code - AKS_STATERESTOREQUICK) / 2, 0);

src/osdep/keyboard.cpp: 					code = AKS_STATESAVEQUICK + v * 2 + ((shift_state || ctrl_state) ? 0 : 1);
```


What am I missing?

**Reply by midwan** [reply]
There have been some updates in WinUAE, that relate to the savestates - I just merged those yesterday, so they can be tested right away. Relevant issue on WinUAE: https://github.com/tonioni/WinUAE/issues/314

From there, I can see that the keyboard shortcuts for quick load/restore are indeed END + 1, and END + SHIFT + 1 (for example, obviously you can use the rest of the keypad numbers as well).

**Reply by midwan** [reply]
Tested here and seems to work:

- End + Shift + Numpad key = Creates save state in slot X (where X = keys 0-9). However, no screenshot is added.
- End + Numpad key = Loads the save state from the X slot (where X = keys 0-9).

I didn't test the period key, that is probably also one of the slots.

**Reply by midwan** [reply]
Update: fixed the no screenshot issue I mentioned above. :)

**Reply by giantclambake** [reply]
Nice... bit late here (long day)...I'll test all this on the morrow when I wake  ;)
